File: topic1/problem12/Solution.py
class RandomizedSet:

    def __init__(self):
        self.vals = list()
        self.idx = dict()

    def insert(self, val: int) -> bool:
        if val in self.idx: return False

        self.vals.append(val)
        self.idx[val] = len(self.vals) -1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.idx: return False

        index = self.idx[val]
        self.vals[index] = self.vals[-1]
        self.idx[self.vals[-1]] = index
        self.vals.pop()
        del self.idx[val]
        
        return True

    def getRandom(self) -> int:
        return random.choice(self.vals)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: topic1/problem12/Solution.java
class RandomizedSet {
    private ArrayList<Integer> vals;
    private Map<Integer, Integer> idx;

    public RandomizedSet() {
        vals = new ArrayList<>();
        idx = new HashMap<>();
    }
    
    public boolean insert(int val) {
        if (idx.containsKey(val)) return false;

        vals.add(val);
        idx.put(val, vals.size() -1);
        return true;
    }
    
    public boolean remove(int val) {
        if (! idx.containsKey(val)) return false;

        int index = idx.get(val);
        vals.set(index, vals.get(vals.size() -1));
        idx.put(vals.get(index), index);
        vals.remove(vals.size() -1);
        idx.remove(val);

        return true;
    }
    
    public int getRandom() {
        Random rand = new Random();
        return vals.get(rand.nextInt(vals.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
